renderdoc/driver/d3d11/d3d11_counters.cpp

```diff
@@ -332,6 +348,73 @@ void D3D11Replay::FillTimersAMD(uint32_t &eventStartID, uint32_t &sampleIndex,
   }
 }
 
+static void FlushGPU(WrappedID3D11Device *pDevice, WrappedID3D11DeviceContext *pImmediateContext)
+{
+  ID3D11Query *pAPIQuery = NULL;
+  D3D11_QUERY_DESC queryDesc;
+  queryDesc.Query = D3D11_QUERY_EVENT;
+  queryDesc.MiscFlags = 0;
+  if(FAILED(pDevice->CreateQuery(&queryDesc, &pAPIQuery)))
+  {
+    return;
+  }
+
+  HRESULT hr;
+  pImmediateContext->Flush();
+  pImmediateContext->End(pAPIQuery);
+  pImmediateContext->Flush();
+  if(S_OK != (hr = pImmediateContext->GetData(pAPIQuery, NULL, 0, D3D11_ASYNC_GETDATA_DONOTFLUSH)))
+  {
+    do
+    {
+      if(FAILED(hr))
+      {
+        return;
+      }
+      ::Sleep(0);    // Give up time slice
+    } while(S_OK !=
+            (hr = pImmediateContext->GetData(pAPIQuery, NULL, 0, D3D11_ASYNC_GETDATA_DONOTFLUSH)));
+  }
+
+  if(pAPIQuery)
+    pAPIQuery->Release();
+}
+
+void D3D11Replay::FillTimersNV(uint32_t &eventStartID, uint32_t &sampleIndex,
+                               vector<uint32_t> &eventIDs, const DrawcallDescription &drawnode)
+{
+  if(drawnode.children.empty())
+    return;
+
+  for(size_t i = 0; i < drawnode.children.size(); i++)
+  {
+    const DrawcallDescription &d = drawnode.children[i];
+
+    FillTimersNV(eventStartID, sampleIndex, eventIDs, drawnode.children[i]);
+
+    if(d.events.empty() || (!(drawnode.children[i].flags & DrawFlags::Drawcall) &&
+                            !(drawnode.children[i].flags & DrawFlags::Dispatch)))
+      continue;
+
+    eventIDs.push_back(d.eventId);
+
+    m_pDevice->ReplayLog(eventStartID, d.eventId, eReplay_WithoutDraw);
+
+    FlushGPU(m_pDevice, m_pImmediateContext);
+
+    m_pNVCounters->BeginSample(sampleIndex);
+
+    m_pDevice->ReplayLog(eventStartID, d.eventId, eReplay_OnlyDraw);
+
+    FlushGPU(m_pDevice, m_pImmediateContext);
+
+    m_pNVCounters->EndSample(sampleIndex);
```

Comment:
It seems strange to flush/serialize the GPU *before* ending the sample. Is this definitely what you want to do? Instead I'd expect to either flush only before starting each sample, or else do `flush -> begin -> replaylog -> end -> flush`.

Reply:
The flush *before* `EndSample` is the necessary thing according to *NVIDIA PerfKit User Guide 4.2.0* (see page 5, section *Experiment Mode*). The flush *before* `BeginSample` is just to make sure that we serialized everything right before draw/dispatch call. Also, similar approach could be found in a sample code (`PerfKit/samples/NvPmSampleDX`) for the *PerfKit*.

Reply:
Fair enough! It seems strange to me so I just wanted to check it was on purpose.